cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java

```diff
@@ -140,4 +157,54 @@ private static URL getURLByClass(ClassLoader classLoader, String className) {
     }
     return ClassLoaders.getClassPathURL(className, resource);
   }
+
+  /**
+   * Starts Py4j gateway server if in executor container and running PySpark.
+   *
+   * @param classLoader the classloader to use for loading classes
+   * @return a {@link Runnable} that calling the {@link Runnable#run()} method will stop the server.
+   */
+  private static Runnable startGatewayServerIfNeeded(ClassLoader classLoader) {
+    Runnable noopRunnable = new Runnable() {
+      @Override
+      public void run() {
+        // no-op
+      }
+    };
+
+    // If we are not running PySpark or
+    // if this process is the AM, no need to start gateway server
+    // The spark execution service uri is always set for the driver (AM) process
+    if (!isPySpark() || System.getenv(SparkRuntimeUtils.CDAP_SPARK_EXECUTION_SERVICE_URI) != null) {
+      return noopRunnable;
+    }
+
+    // Otherwise start the gateway server using reflection. Also write the port number to a local file
+    Path portFile = Paths.get("cdap.py4j.gateway.port.txt");
+    try {
+      final Object server = classLoader.loadClass(SparkPythonUtil.class.getName())
+        .getMethod("startPy4jGateway", Path.class).invoke(null, portFile);
+
+      LOG.info("Py4j GatewayServer started, listening at port {}",
+               new String(Files.readAllBytes(portFile), StandardCharsets.UTF_8));
+
+      return new Runnable() {
+        @Override
+        public void run() {
+          try {
+            server.getClass().getMethod("shutdown").invoke(server);
+          } catch (Exception e) {
+            LOG.warn("Failed to shutdown Py4j GatewayServer", e);
+          }
+        }
+      };
+    } catch (Exception e) {
+      LOG.warn("Failed to start Py4j GatewayServer. No CDAP functionality will be available in executor", e);
+      return noopRunnable;
+    }
+  }
+
+  private static boolean isPySpark() {
+    return System.getenv("PYTHONPATH") != null;
```

Comment:
someone might've modified `app.program.jvm.opts` to include "PYTHONPATH", in case they're trying to execute some python transform or script transform from some stage, right? Wouldn't that mess this check up?

Reply:
This is environment, won't be affected by jvm opts

Reply:
I mean - is there any way user could control environment of application containers?

Reply:
This is an env setup by Spark. If user is running with PySpark, there is no point to mess it up, as things won't work. If user is not running with PySpark, and manually setting this, it won't affect CDAP much. I think we can deal with it if it really becomes a problem (in non PySpark case only).

Reply:
"If user is not running with PySpark, and manually setting this, it won't affect CDAP much."
Ok, cool.